src/wayland/seat/keyboard.rs

```diff
@@ -0,0 +1,201 @@
+    ///
+    /// This should be done first, before anything else is done with this keyboard.
+    pub(crate) fn new_kbd(&self, kbd: WlKeyboard) {
+        trace!(self.arc.logger, "Sending keymap to client");
+
+        // prepare a tempfile with the keymap, to send it to the client
+        let ret = self.arc.keymap.with_fd(kbd.version() >= 7, |fd, size| {
+            kbd.keymap(KeymapFormat::XkbV1, fd, size as u32);
+        });
+
+        if let Err(e) = ret {
+            warn!(self.arc.logger,
+                "Failed write keymap to client in a tempfile";
+                "err" => format!("{:?}", e)
+            );
+            return;
+        };
+
+        let guard = self.arc.internal.lock().unwrap();
+        if kbd.version() >= 4 {
+            kbd.repeat_info(guard.repeat_rate, guard.repeat_delay);
+        }
+        if let Some((focused, serial)) = guard.focus.as_ref() {
+            if focused.same_client_as(&kbd.id()) {
+                let serialized = guard.mods_state.serialized;
+                let keys = serialize_pressed_keys(guard.pressed_keys.clone());
+                kbd.enter((*serial).into(), focused.wl_surface().unwrap(), keys);
+                // Modifiers must be send after enter event.
+                kbd.modifiers(
+                    (*serial).into(),
+                    serialized.depressed,
+                    serialized.latched,
+                    serialized.locked,
+                    serialized.layout_locked,
+                );
+            }
+        }
+        self.arc.known_kbds.lock().unwrap().push(kbd);
+    }
+}
+
+/// User data for keyboard
+#[derive(Debug)]
+pub struct KeyboardUserData<D: SeatHandler> {
+    pub(crate) handle: Option<KeyboardHandle<D>>,
+}
+
+impl<D> Dispatch<WlKeyboard, KeyboardUserData<D>, D> for SeatState<D>
+where
+    D: 'static + Dispatch<WlKeyboard, KeyboardUserData<D>>,
+    D: SeatHandler,
+{
+    fn request(
+        _state: &mut D,
+        _client: &wayland_server::Client,
+        _resource: &WlKeyboard,
+        _request: wl_keyboard::Request,
+        _data: &KeyboardUserData<D>,
+        _dhandle: &DisplayHandle,
+        _data_init: &mut wayland_server::DataInit<'_, D>,
+    ) {
+    }
+
+    fn destroyed(_state: &mut D, _client_id: ClientId, object_id: ObjectId, data: &KeyboardUserData<D>) {
+        if let Some(ref handle) = data.handle {
+            handle
+                .arc
+                .known_kbds
+                .lock()
+                .unwrap()
+                .retain(|k| k.id() != object_id)
+        }
+    }
+}
+
+fn with_focused_kbds<D: SeatHandler + 'static>(
+    seat: &Seat<D>,
+    surface: &WlSurface,
+    mut f: impl FnMut(WlKeyboard),
+) {
+    if let Some(keyboard) = seat.get_keyboard() {
+        let inner = keyboard.arc.known_kbds.lock().unwrap();
+        for kbd in &*inner {
+            if kbd.id().same_client_as(&surface.id()) {
+                f(kbd.clone())
+            }
```

Comment:
Could we rename this to `for_each_focused_kbd`?

Reply:
done